### PR TITLE
TST: migrate image comparison tests from nose to pytest-mpl (1/2)

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary
This is a self-contained subset of #4118, targetting only tests that can be run with fake datasets and don't require big data infrastructure. These correspond exactly to the GHA answer-tests job.
partially addressed #3735

TODO:
- [x] check that the answer-tests GHA job is reduced to 0 tests
- [x] polish tests (reduce repetition using `pytest.mark.parametrize` ?)
- [x] make sure all tests have a new baseline (edit: not planning to follow through on this for mesh rendering tests requiring pyembree, because it's really not trivial to install either locally or on CI for me atm)
- [ ] document development workflows (how to update baseline; how to create new image tests)
- [x] build any missing supporting architecture for the newly translated tests (namely, create a separate org-level repo to contain reference image files)

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Making progress, merging #4125 and #4127 would help a lot with a couple issues that are tiresome to work around locally.

**Comment by neutrinoceros** [comment]
failures on windows are unrelated (#4128), but I learned something important from the image tests failure:
I was attempting to use pytest-mpl in "hybrid mode" (prioritise hash comparison, then go to image comparison in case of failure), but my baselines (both hashes and images) were generated on my laptop, which gives slightly different results from what is generated in CI; the differences in images are low enough to pass tests but (for good reasons) not differences in *hashes*. So I guess we can't use the hybrid mode (or hash comparison) as it completely negates having a non-zero RMS tolerance.

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
Currently experiencing hangs on Jenkins which are specific to this PR. This is the last outputs that are logged before the job hangs.

```
/home/fido/workspace/yt_py38_git/yt/visualization/volume_rendering/render_source.py:85: UserWarning: Failed to switch to embree raytracing engine. The following error was raised:
cannot import name 'mesh_construction' from 'yt.utilities.lib.embree_mesh' (/home/fido/workspace/yt_py38_git/yt/utilities/lib/embree_mesh/__init__.py)
  warnings.warn(
Traceback (most recent call last):
  File "tests/nose_runner.py", line 121, in <module>
    for job in generate_tasks_input():
  File "tests/nose_runner.py", line 83, in generate_tasks_input
    for test in list(tests["other_tests"].keys()):
KeyError: 'other_tests'
```

Assuming I didn't just break the global state of the server somehow, I'm going to try and fix the problems I'm seeing here

**Comment by neutrinoceros** [comment]
one test is currently unstable against matplotlib 3.2, that was only ever tested with mpl 3.5 and 3.6
I'll dig in to see if this can be mitigated or if the test should simply be skipped on older versions

**Comment by neutrinoceros** [comment]
The one thing I originally planned to do but didn't is documentation. I still plan to do it but currently there's nothing to replace (the existing workflow, which requires updating the answer-store, is extremely similar to the one needed after this PR), so it seems fine. In all cases this PR will require follow up to:
- remove answer-tests GHA job (I'm leaving it here, even if it runs 0 tests, because it's _required_ by auto-merge)
- require the replacement job instead
- drop the answer-store submodule

In the mean time, I now consider this stable and ready for review

**Comment by neutrinoceros** [comment]
force-pushed to resolve conflicts

**Comment by matthewturk** [comment]
OK, so I have read this, and I think it has a large number of very-difficult-to-review components.  I see a few options:

- Split the PR into one that removes *no* code and just adds the new image testing *base* functions and one that does the switch
- Check and confirm that we have a) the same number of tests b) the same results if we intentionally break the tests.  (this is my preferred)  For that, you'd have to show that the image comparison tests break if we change the unit representations, or if we add an offset to the field values so that they produce different image buffers, for instance.

**Comment by neutrinoceros** [comment]
> - Split the PR into one that removes no code and just adds the new image testing base functions and one that does the switch

I think that would be my preferred track, but can you clarify what you mean here by "base" functions ?

> Check and confirm that we have a) the same number of tests b) the same results if we intentionally break the tests. (this is my preferred) For that, you'd have to show that the image comparison tests break if we change the unit representations, or if we add an offset to the field values so that they produce different image buffers, for instance.

- I honestly don't know how to do a)
- Introducing errors on purpose to see tests fail can be done from the reviewer standpoint, so what is there for me to do on this step exactly ? Should I provide an explicit "how-to test this branch" here ?

**Comment by neutrinoceros** [comment]
rebased on main with no change to refresh CI. Unless anything bad happens I'll also bump the new test job to Python 3.11

File: .github/workflows/build-test.yaml
Comment (matthewturk) [inline .github/workflows/build-test.yaml:132]: Just so I understand, this will only run in case of failure, and in that particular case it will upload the artifacts?

File: conftest.py
Comment (matthewturk) [inline conftest.py:212]: Can you explain this?  Why is tolerance 0.5?  What does this do?

File: nose_unit.cfg
Comment (matthewturk) [inline nose_unit.cfg:9]: Do you think we'll ever be able to break this into multiple lines? :)

File: yt/fields/tests/test_field_access.py
Comment (matthewturk) [inline yt/fields/tests/test_field_access.py:31]: Just so I am clear, this PR changes it to use the current API, but the old API is still currently in place, right?  This is just keeping up to date with a deprecation?

File: yt/utilities/answer_testing/framework.py
Comment (matthewturk) [inline yt/utilities/answer_testing/framework.py:1043]: @brittonsmith just wanted to check that this does *not* get used in enzo or grackle, right?

File: yt/utilities/answer_testing/framework.py
Comment (matthewturk) [inline yt/utilities/answer_testing/framework.py:1085]: big QOL improvement here IMO

**Review by matthewturk — APPROVED** [review]
So I'm marking approve, because I think we need to eventually pull the ripcord and go on this.  I think what we need to do is basically convince ourselves we're not missing anything major, and then just make the switch.

Do you have a sense of if this is any faster?

File: nose_unit.cfg
Comment (neutrinoceros) [reply nose_unit.cfg:9]: I wish. This one line generates so many conflicts that I have a nickname for it (I call it "the blockchain") ... I think I once tried splitting the argument and found that it wasn't supported. Supposedly it's possible to pass the equivalent command line argument more than once to combine regexes, but AFAIU this doesn't work with static configuration files. So if we replace this line by command line arguments, we may be able to escape conflict hell, but that'd be an entire mini project in itself.

File: .github/workflows/build-test.yaml
Comment (neutrinoceros) [reply .github/workflows/build-test.yaml:132]: Not exactly. Note that steps that do the uploads are conditioned with `always()`, so we try to upload as many artefacts as we produced. The "new baseline" artifact may not exist because it's only generated when needed, but the html report should always be present.

File: conftest.py
Comment (neutrinoceros) [reply conftest.py:212]: Here I'm providing arguments to be passed by default to `@pytest.marker.mpl_image_compare` so they don't have to be repeated.
I went with 0.5, which is [stricter than the default](https://github.com/matplotlib/pytest-mpl#tolerance), because I figured we might want to be extra careful at first since this is a big leap of a transition.

File: yt/fields/tests/test_field_access.py
Comment (neutrinoceros) [reply yt/fields/tests/test_field_access.py:31]: the `_setup_plots` method isn't actually deprecated, I just think it's best to avoid calling private methods directly from outside a class (which is the whole reason I added a public equivalent in this case).
But yes, it's just a no-op change toward stable API

**Comment by neutrinoceros** [comment]
>  I think what we need to do is basically convince ourselves we're not missing anything major

The only major thing that's missing IMO is dev docs. Because the current process for updating image baselines is _already_ not documented and also extremely similar to the one I settled for here, I figured this could probably wait for a follow up PR, but the more this stays waiting, he harder it'll be for me to write it after the fact. I should be able to prioritise this task this week. What would be easier to review ? another 200 lines within this PR, or a brand new one after this one is merged ?

> Do you have a sense of if this is any faster?

I don't think it is. In fact it currently takes *longer* on GHA

pytest-mpl:
```
==== 143 passed, 30 skipped, 915 deselected, 1 warning in 229.02s (0:03:49) ====
```

nose
```
----------------------------------------------------------------------
[146](https://github.com/yt-project/yt/actions/runs/4231076474/jobs/7349149421#step:8:147)
Ran 105 tests in 153.973s
```

I assume the reason is that more tests cases are actually run (on nose, we didn't actually cover cartopy integration and mesh rendering !)

File: nose_unit.cfg
Comment (neutrinoceros) [reply nose_unit.cfg:9]: I'm addressing this in #4346

**Comment by neutrinoceros** [comment]
updated following #4346

File: yt/utilities/answer_testing/framework.py
Comment (matthewturk) [reply yt/utilities/answer_testing/framework.py:1043]: I don't *think* we do.  But that's why I'm asking!

Comment (brittonsmith) [reply yt/utilities/answer_testing/framework.py:1043]: Grackle doesn't use either of these classes. Enzo uses `AnswerTestingTest`, but not `GenericImagetest`.

Comment (neutrinoceros) [reply yt/utilities/answer_testing/framework.py:1043]: `AnswerTestingTest` isn't removed in this PR, so it should be good !

Comment (neutrinoceros) [reply yt/utilities/answer_testing/framework.py:1043]: @matthewturk any chance we might move forward with this soon ?

**Comment by matthewturk** [comment]
Yes, I actually don't think we need to wait any longer.

**Review by matthewturk — APPROVED** [review]
Do we think this is done?  @Xarthisius have you had a chance to look it over?

**Comment by neutrinoceros** [comment]
trying our luck again: ping @Xarthisius

**Comment by neutrinoceros** [comment]
I'd really like to move forward with this one, I consider it blocking for many image-related changes I would like to get going.

**Comment by Xarthisius** [comment]
I'm not sure why I was pinged on this PR. If you're waiting for my review just go ahead.

**Pull request merged**
